core/data_manager.py

```diff
@@ -311,12 +311,40 @@ def update_application(self, app: dict) -> None:
         try:
             cur = self.db_conn.cursor()
             cur.execute('''REPLACE INTO applications (id, name, guid, user_id) 
-                                       VALUES (?, ?, ?, ?)''', 
+                                        VALUES (?, ?, ?, ?) ''', 
                         [app['id'], app['name'], app['guid'], app['user_id']])
             self.db_conn.commit()
         except Error as e:
             print('[Error] in update_application:')
             print(e)
+
+    def save_application(self, app: dict) -> None:
+        try:
+            cur = self.db_conn.cursor()
+            if(app['id'] == -1):
+                cur.execute('''SELECT IFNULL(MAX(id), 0) + 1 FROM applications;''')
+                app['id'] = cur.fetchone()[0]
+            print(app)
+            cur.execute('''INSERT INTO applications (id, name, user_id) 
+                                    VALUES (?, ?, ?) ON CONFLICT (id) DO UPDATE SET
+                                    name=excluded.name''', 
+                        [app['id'], app['name'], app['user_id']])
+            self.db_conn.commit()
+        except Error as e:
+            print('[Error] in save_application:')
+            print(e)
+
+    def delete_application(self, app: dict) -> bool:
+        try:
+            cur = self.db_conn.cursor()
+            cur.execute('''PRAGMA foreign_keys=on;''')
```

Comment:
This part I don't know. Please add comments to explain why the `PRAGMA foreign_keys=on;` is needed for this query.

Reply:
Sorry for the confusion, just double checked that PRAGMA foreign_keys=on is not needed for the 'applications' query, so I have deleted it.